docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,286 @@
+- Superior caching
+
+What has changed?
+-----------------
+
+- ``Global Bears`` are now called ``Project Bears`` and ``Local Bears`` are
+  now known as ``File Bears``. Both the ``ProjectBear`` and the ``FileBear``
+  classes inherit from the base class ``coalib.core.Bear``.
+- There is no need to pass ``HiddenResults`` between different bears (made
+  possible by the new dependency management system) to hide results to the
+  result callback. Only the results that were explicitly requested by passing
+  the needed bears are passed now. There is no need to pass instances of
+  ``Result`` anymore, any python object will do.
+- The former ``run`` function that was inherited by all the bears to run code
+  analysis is now replaced by the ``analyze`` function.
+
+Easier Interface
+----------------
+
+Running a coala session with the NextGen-Core is now as easy as
+``core.run(bears)``. The ``run`` method takes the following arguments ``bears``,
+``result_callback``, ``cache`` and ``executor`` (the last two are ``None`` by
+default) and initiates a coala session.
+
+* The ``bears`` argument contains the list of bears to be run for a coala
+  session.
+* The ``result_callback`` is a function that is called when the results from a
+  bear run are available.
+* The ``cache`` argument if provided enables caching and runs the session using
+  the cache provided to store the bear results.
+* The ``executor`` argument is used to provide a custom executor in which the
+  passed bears are to be run which is closed after the core is closed. If this
+  argument is not provided then ``ProcessPoolExecutor`` is used using as many
+  processes as cores available on the system.
+
+Official support for virtual files
+----------------------------------
+
+IDEs like IntelliJ use a virtual files to represent files in a filesystem (VFS)
+and perform operations on them. Hence NextGen-Core provides official support
+for virtual files, the bears have to point to the right file data objects when
+run, whether they are real files or virtual ones. This makes coala easier to
+integrate with IDEs since the contents reside in the IDE memory which is faster
+instead of reloading files from the disk.
+
+Task Objects
+------------
+
+These are the representation of the tasks performed by the bears. Structure-wise
+they are a tuple containing the tuple of positional arguments and a dict of
+keyword arguments to the bears. The task objects are instrumental to the caching
+mechanism as their hash values are stored in the cache along with the bear
+results and during each coala run the core looks for identical task objects in
+the cache, and fetches their corresponding results if found without even
+executing those tasks thereby providing a performance boost. Below you can see
+an example of two task objects.
+
+::
+
+    [
+        [(file, filename), {'filename_prefix': "", 'filename_suffix': ""}],
+        [(filename, file), {'preferred_quotation': ""}],
+    ]
```

Comment:
I think we should put this into some very simple bear implementation, what do you think?

Reply:
Well, I was going for a minimalistic approach to just show what task objects are in terms of bears parameters and show them together in the form of tuples and dicts.
I think the name of the task objects are pretty self-explanatory for a bear writer to know where they will go inside a bear and which ones are positional and which ones are keyword args.
But if you still want a bear implementation let me know I'll add one along with this, leaving this unchanged for now. :smile:

Reply:
I think you should add a simple bear implementation which incorporates this. It will make the documentation look more complete imo.